### PR TITLE
Feature/comment system backend

### PR DESCRIPTION
**Pull request by Letaguator**
Current maximum supported depth is 5 (including root level) for fetching comments and it's child comments. We may wanna extend this. Think the JSON parser supports up to 32 in depth. 

It is technically possible to reply to the leaf node comments, but one would be unable to fetch this level. We may wanna do something about this, or just enforce it client side by denying replying to level 5.

### CODE REVIEW

**Comment by AndreasJJ** [comment]
Any reason why its limited to 5 and 32? https://docs.microsoft.com/en-us/dotnet/standard/serialization/system-text-json-migrate-from-newtonsoft-how-to#maximum-depth from those two functions/libraries/thingies it seems like there's no limit/configurable. What function/library/thingy do we use?

File: backend/Controllers/CommentController.cs [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+using OpenData.Resources;
+using OpenData.Domain.Models;
+
+namespace OpenData.Controllers
+{
+    [Authorize]
+    [Route("/api/[controller]")]
+    public class CommentController : Controller
+    {
+        private readonly ICommentService commentService;
```
Comment (petterroea) [inline backend/Controllers/CommentController.cs]: Reccomend using the `_` prefix for private variables - this is C# standard, and lets us avoid using `this.`

File: backend/Controllers/ExperiencePostController.cs [hunk 2]
```diff
@@ -92,5 +98,33 @@ public async Task<IActionResult> PostAsync([FromBody] Tag newTag, string uuid)
...
 			return Ok(res);
 		}
+
+		/// <summary>
+		/// Used to PUT comments related to a given Metadata.
+		/// </summary>
+		/// <param name="experiencePostGuid"></param>
+		/// <param name="newComment">New comment to be added.</param>
+		/// <returns>The comment if it was added successfully</returns>
+		[HttpPut("{metadataGuid}/comments")]
```
Comment (petterroea) [inline backend/Controllers/ExperiencePostController.cs]: Shouldn't this be `experiencePostGuid`?

File: backend/Persistence/Repositories/CommentRepository.cs [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+
+        public async Task<IEnumerable<Comment>> FetchCommentsForExperiencePostAsync(Guid experiencePostUuid)
+        {
+            return await _context
+                .ExperiencePostCommentMappings
+                .Include(m => m.Comment)
+                .ThenInclude(m => m.ChildComments)
+                .ThenInclude(m => m.ChildComments)
+                .ThenInclude(m => m.ChildComments)
+                .ThenInclude(m => m.ChildComments)
```
Comment (petterroea) [inline backend/Persistence/Repositories/CommentRepository.cs]: What does this do? Why 4 lines that does the same?

File: backend/Persistence/Repositories/MetadataRepository.cs [hunk 4]
```diff
@@ -6,6 +6,7 @@
 using OpenData.Persistence.Contexts;
 using System.Web;
 using System;
+using System.Linq;
```
Comment (petterroea) [inline backend/Persistence/Repositories/MetadataRepository.cs]: ?

**Comment by Letaguator** [comment]
> Any reason why its limited to 5 and 32? https://docs.microsoft.com/en-us/dotnet/standard/serialization/system-text-json-migrate-from-newtonsoft-how-to#maximum-depth from those two functions/libraries/thingies it seems like there's no limit/configurable. What function/library/thingy do we use?

The reason for the 5 limit is since I have 4 "Includes" in the the query. In the sense that I have instructed it to tech child comments for comments 4 times. The 32 limit may be incorrect, but I think that is what it said when I had an error related to a cycle. Something about "no cycle and no-depth larger than 32", may remember the exact number incorrectly.

File: backend/Persistence/Repositories/MetadataRepository.cs (hunk 4 above)
Comment (Letaguator) [reply backend/Persistence/Repositories/MetadataRepository.cs]: Ahhh, this is from when the comment had functions in the MetadataController.

**Comment by petterroea** [comment]
So comment trees deeper than 5 isn't supported?

File: backend/Controllers/ExperiencePostController.cs (hunk 2 above)
Comment (Letaguator) [reply backend/Controllers/ExperiencePostController.cs]: Correct, will fix it asap! :)

File: backend/Persistence/Repositories/CommentRepository.cs (hunk 3 above)
Comment (Letaguator) [reply backend/Persistence/Repositories/CommentRepository.cs]: It fetches the child comments with a max depth of 5 (including root).

File: backend/Persistence/Repositories/MetadataRepository.cs (hunk 4 above)
Comment (Letaguator) [reply backend/Persistence/Repositories/MetadataRepository.cs]: Fixed

**Comment by Letaguator** [comment]
We can have deeper than 5, at least up to 32, maybe even higher.

File: backend/Persistence/Repositories/CommentRepository.cs (hunk 3 above)
Comment (petterroea) [reply backend/Persistence/Repositories/CommentRepository.cs]: Why only max depth of 5

**Comment by Letaguator** [comment]
It is connected to two stories so thought we might as well. If you have another way you wanna do it, please suggest it.

File: backend/Persistence/Repositories/CommentRepository.cs (hunk 3 above)
Comment (Letaguator) [reply backend/Persistence/Repositories/CommentRepository.cs]: Can add more .ThenInclude if we want to increase the depth.

**Comment by Letaguator** [comment]
Since we wanna have it large 32 seems like a too random number to stop at. Will therefore look at increasing it to infinity. This will require frontend to explicitly say when they wanna fetch five more in a give comment tree. Will be something like GET /api/comment/{uuid}/childcomments.

**Comment by Letaguator** [comment]
Sounds good @AndreasJJ and @MittBrukernavn?

**Comment by Letaguator** [comment]
Have redesigned the system to support infinite depth.

**Pull request merged**